### DISCUSSION TITLE
Confused by CLAP_EVENT_NOTE_EXPRESSION (e.g. Pitchbend)

### DISCUSSION THREAD

**Discussion by sagantech** [free-audio/clap#203] [Q&A]
So I am wondering the preferred method of a host to translate straight MIDI events (say from a standard MIDI controller) to CLAP_NOTE_DIALECT_CLAP.  I am most concerned about the bend wheel and the modulation wheel. I notice that some plug-ins respond to CLAP_NOTE_EXPRESSION_TUNING for pitch bending but I assume that is not the preferred method. Also it does not look like many plug-ins respond to CLAP_NOTE_EXPRESSION_VIBRATO if I send that for modulation (controller 1).

Any help greatly appreciated.

**Comment by robbert-vdh** [comment]
If a plugin supports it, `CLAP_NOTE_EXPRESSION_TUNING` is the preferred method for pitch expressions since it has a defined range of -120 to +120 semitones. In contrast, MIDI pitch bends are untyped. That also means you cannot translate MIDI pitch bend into tuning expressions without defining the a pitch bend range first. If they're MIDI pitch bends and the plugin supports MIDI events, you should just pass them through as is using `CLAP_EVENT_MIDI`.

**Reply by sagantech** [reply]
Thanks for the quick reply.
Some plug-ins do not support CLAP_EVENT_MIDI. My host **does** know what range of pitch bend the user desires. 
What about modulation?

**Reply by robbert-vdh** [reply]
You can translate the pitch bend to note expressions then. I'd just add a per-plugin toggle somewhere to let the user choose between MIDI messages and CLAP note expressions if the plugin supports both note event formats.

**Reply by sagantech** [reply]
> translate the pitch bend to note expressions

Do you mean by using CLAP_NOTE_EXPRESSION_TUNING or is there some other way?

**Reply by robbert-vdh** [reply]
Yes, `CLAP_NOTE_EXPRESSION_TUNING`.

**Reply by sagantech** [reply]
I already have a way to select the MIDI dialect per plug-in. Here are the explicit issues I am facing and admittedly they are minor:

1. Apricot does not respond to modulation wheel (CLAP_NOTE_EXPRESSION_VIBRATO ) and does not support CLAP_EVENT_MIDI. Works fine with CLAP_NOTE_EXPRESSION_TUNING
2. The just released HIVE works fine with CLAP_EVENT_MIDI but in CLAP_NOTE_DIALECT_CLAP does not respond to CLAP_NOTE_EXPRESSION_TUNING or CLAP_NOTE_EXPRESSION_VIBRATO. The GUI pitch bend wheel works fine in this dialect.
